File: ios/StashKit/Sources/StashKit/Chat/ChatSessions.swift
import Foundation

/// Port of the web's `src/utils/chatSessions.ts` (spec:
/// `docs/superpowers/specs/2026-08-27-chat-sessions-design.md`): a conversation is a burst of
/// activity — 3+ hours of silence starts a new one. Pure logic only; the DB side (the
/// `last_message_at` trigger and `list_conversations` RPC) is already deployed and shared with
/// the web client.
public enum ChatSessions {
    /// `SESSION_GAP_MS`, in seconds.
    public static let sessionGap: TimeInterval = 3 * 60 * 60

    public struct Candidate: Equatable, Sendable {
        public let id: UUID
        public let title: String?
        public let lastMessageAt: Date?

        public init(id: UUID, title: String?, lastMessageAt: Date?) {
            self.id = id
            self.title = title
            self.lastMessageAt = lastMessageAt
        }
    }

    public enum Target: Equatable, Sendable {
        case continueSession(id: UUID, title: String?)
        case new
    }

    /// `resolveSessionTarget` — continue the latest conversation iff its last message is under
    /// the gap old; anything else (no conversation, no timestamp, stale) starts fresh.
    public static func resolveTarget(latest: Candidate?, now: Date) -> Target {
        guard let latest, let lastMessageAt = latest.lastMessageAt else { return .new }
        return now.timeIntervalSince(lastMessageAt) < sessionGap
            ? .continueSession(id: latest.id, title: latest.title)
            : .new
    }

    /// `bucketConversations`' label rule, for one row: Today / Yesterday / This week (Monday
    /// start, matching date-fns `weekStartsOn: 1`) / month name (year appended when not the
    /// current year). Rows arrive newest-first from the RPC, so emitting labels in row order
    /// groups correctly with one pass.
    public static func bucketLabel(for date: Date, now: Date, calendar base: Calendar = .current) -> String {
        var calendar = base
        calendar.firstWeekday = 2
        let days = calendar.dateComponents([.day], from: calendar.startOfDay(for: date),
                                            to: calendar.startOfDay(for: now)).day ?? 0
        if days <= 0 { return "Today" }
        if days == 1 { return "Yesterday" }
        if calendar.isDate(date, equalTo: now, toGranularity: .weekOfYear) { return "This week" }
        let formatter = DateFormatter()
        formatter.calendar = calendar
        formatter.locale = Locale(identifier: "en_US_POSIX")
        let sameYear = calendar.isDate(date, equalTo: now, toGranularity: .year)
        formatter.dateFormat = sameYear ? "MMMM" : "MMMM yyyy"
        return formatter.string(from: date)
    }

    /// PostgREST renders `timestamptz` as ISO8601, with fractional seconds for most rows but
    /// without them for values that landed on an exact second — accept both.
    public static func parseTimestamp(_ raw: String?) -> Date? {
        guard let raw else { return nil }
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = fractional.date(from: raw) { return date }
        let plain = ISO8601DateFormatter()
        plain.formatOptions = [.withInternetDateTime]
        return plain.date(from: raw)
    }
}

/// One row of `list_conversations(search_text, page_limit, page_offset)` (migration
/// `20260828100000`): search matches title OR any message content, pages clamp 1–100 server-side,
/// every row carries the filtered `total_count`.
public struct ConversationListRow: Identifiable, Equatable, Sendable {
    public let id: UUID
    public let title: String?
    public let lastMessageAt: Date
    public let messageCount: Int
    public let preview: String?
    public let totalCount: Int

    public init(id: UUID, title: String?, lastMessageAt: Date, messageCount: Int,
                preview: String?, totalCount: Int) {
        self.id = id
        self.title = title
        self.lastMessageAt = lastMessageAt
        self.messageCount = messageCount
        self.preview = preview
        self.totalCount = totalCount
    }
}

File: src/utils/chatSessions.ts
import { differenceInCalendarDays, format, isSameWeek } from 'date-fns';

// A conversation is a burst of activity: 3+ hours of silence starts a new one.
export const SESSION_GAP_MS = 3 * 60 * 60 * 1000;

export interface SessionCandidate {
  id: string;
  title: string | null;
  last_message_at: string | null;
}

export function resolveSessionTarget(
  latest: SessionCandidate | null,
  now: Date
): { kind: 'continue'; id: string; title: string | null } | { kind: 'new' } {
  if (!latest?.last_message_at) return { kind: 'new' };
  const elapsed = now.getTime() - new Date(latest.last_message_at).getTime();
  return elapsed < SESSION_GAP_MS
    ? { kind: 'continue', id: latest.id, title: latest.title }
    : { kind: 'new' };
}

export interface ConversationListRow {
  id: string;
  title: string | null;
  last_message_at: string;
  message_count: number;
  preview: string | null;
}

export interface ConversationBucket {
  label: string;
  rows: ConversationListRow[];
}

// Rows arrive newest-first (list_conversations orders by last_message_at desc),
// so buckets emit in display order with one pass.
export function bucketConversations(
  rows: ConversationListRow[],
  now: Date
): ConversationBucket[] {
  const buckets: ConversationBucket[] = [];
  for (const r of rows) {
    const d = new Date(r.last_message_at);
    const days = differenceInCalendarDays(now, d);
    let label: string;
    if (days <= 0) label = 'Today';
    else if (days === 1) label = 'Yesterday';
    else if (isSameWeek(d, now, { weekStartsOn: 1 })) label = 'This week';
    else label = d.getFullYear() === now.getFullYear() ? format(d, 'MMMM') : format(d, 'MMMM yyyy');
    const last = buckets[buckets.length - 1];
    if (last && last.label === label) last.rows.push(r);
    else buckets.push({ label, rows: [r] });
  }
  return buckets;
}
